Userland/Applications/Dictionary/DictionaryModel.cpp

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright (c) 2021, Robin Allen <r@foon.uk>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include "DictionaryModel.h"
+
+namespace Dictionary {
+
+int DictionaryModel::row_count(GUI::ModelIndex const& index) const
+{
+    if (!index.is_valid()) {
+        return m_last_index - m_first_index;
+    }
+    return 0;
+}
+
+int DictionaryModel::column_count(GUI::ModelIndex const&) const
+{
+    return 1;
+}
+
+GUI::Variant DictionaryModel::data(GUI::ModelIndex const& index, GUI::ModelRole role) const
+{
+    if (index.parent().is_valid())
+        return {};
+
+    if (role == GUI::ModelRole::Display || role == GUI::ModelRole::Sort || role == GUI::ModelRole::Search) {
+
+        int row = index.row();
+
+        if (0 <= row && row < (m_last_index - m_first_index)) {
+            return m_dictionary.word_at_index(m_first_index + row);
+        }
+    }
+
+    return {};
+}
+
+DictionaryModel::DictionaryModel()
+    : m_dictionary("/res/dictionaries/wordnet.bin")
+{
+    m_first_index = 0;
+    m_last_index = m_dictionary.count_words();
+}
+
+void DictionaryModel::set_query(String query)
+{
+    int num_words = m_dictionary.count_words();
+
+    if (query == "") {
+        set_range(0, num_words);
+        m_query = query;
```

Comment:
since you're saving the caller's query here, should `move()` it into place.

```c++
        m_query = move(query);
```

Reply:
Done.